src/main/io/gps.c

```diff
@@ -573,110 +848,249 @@ static void ubloxSetMessageRate(uint8_t messageClass, uint8_t messageID, uint8_t
+
+    offset += ubloxAddValSet(&tx_buffer, CFG_MSGOUT_NMEA_ID_GLL_I2C, payload, offset);
+    offset += ubloxAddValSet(&tx_buffer, CFG_MSGOUT_NMEA_ID_GLL_SPI, payload, offset);
+    offset += ubloxAddValSet(&tx_buffer, CFG_MSGOUT_NMEA_ID_GLL_UART1, payload, offset);
+
+    offset += ubloxAddValSet(&tx_buffer, CFG_MSGOUT_NMEA_ID_GSA_I2C, payload, offset);
+    offset += ubloxAddValSet(&tx_buffer, CFG_MSGOUT_NMEA_ID_GSA_SPI, payload, offset);
+    offset += ubloxAddValSet(&tx_buffer, CFG_MSGOUT_NMEA_ID_GSA_UART1, payload, offset);
+
+    offset += ubloxAddValSet(&tx_buffer, CFG_MSGOUT_NMEA_ID_RMC_I2C, payload, offset);
+    offset += ubloxAddValSet(&tx_buffer, CFG_MSGOUT_NMEA_ID_RMC_SPI, payload, offset);
+    offset += ubloxAddValSet(&tx_buffer, CFG_MSGOUT_NMEA_ID_RMC_UART1, payload, offset);
+
+    ubloxSendConfigMessage(&tx_buffer, MSG_CFG_VALSET, offsetof(ubxCfgValSet_t, cfgData) + offset, true);
+}
+
+static void ubloxSetNavRate(uint16_t measRate, uint16_t navRate, uint8_t timeRef)
+{
     uint16_t measRateMilliseconds = 1000 / measRate;
-    // Testing has  revealed this is the max rate common modules can achieve
-    if (measRateMilliseconds < 53) measRateMilliseconds = 53;
 
-    tx_buffer.payload.cfg_rate.measRate = measRateMilliseconds;
-    tx_buffer.payload.cfg_rate.navRate = navRate;
-    tx_buffer.payload.cfg_rate.timeRef = timeRef;
-    ubloxSendConfigMessage(&tx_buffer, MSG_CFG_RATE, sizeof(ubxCfgRate_t));
+    ubxMessage_t tx_buffer;
+    if (gpsData.ubloxM9orAbove) {
+        uint8_t offset = 0;
+        uint8_t payload[2];
+        payload[0] = (uint8_t)(measRateMilliseconds >> (8 * 0));
+        payload[1] = (uint8_t)(measRateMilliseconds >> (8 * 1));
+        //rate meas is U2
+        offset = ubloxValSet(&tx_buffer, CFG_RATE_MEAS, payload, UBX_VAL_LAYER_RAM);
+
+        payload[0] = (uint8_t)(navRate >> (8 * 0));
+        payload[1] = (uint8_t)(navRate >> (8 * 1));
+        //rate nav is U2
+        offset += ubloxAddValSet(&tx_buffer, CFG_RATE_NAV, payload, 6);
+
+        payload[0] = timeRef;
+        //rate timeref is E1 = U1
+        offset += ubloxAddValSet(&tx_buffer, CFG_RATE_TIMEREF, payload, 12);
+
+        ubloxSendConfigMessage(&tx_buffer, MSG_CFG_VALSET, offsetof(ubxCfgValSet_t, cfgData) + offset, false);
+    } else {
+        tx_buffer.payload.cfg_rate.measRate = measRateMilliseconds;
+        tx_buffer.payload.cfg_rate.navRate = navRate;
+        tx_buffer.payload.cfg_rate.timeRef = timeRef;
+        ubloxSendConfigMessage(&tx_buffer, MSG_CFG_RATE, sizeof(ubxCfgRate_t), true);
+    }
 }
 
 static void ubloxSetSbas(void)
 {
     ubxMessage_t tx_buffer;
 
-    //NOTE: default ublox config for sbas mode is: UBLOX_MODE_ENABLED, test is disabled
-    tx_buffer.payload.cfg_sbas.mode = UBLOX_MODE_TEST;
-    if (gpsConfig()->sbasMode != SBAS_NONE) {
-        tx_buffer.payload.cfg_sbas.mode |= UBLOX_MODE_ENABLED;
-    }
+    if (gpsData.ubloxM9orAbove) {
```

Comment:
As above, I strongly recommend splitting up functions/files based on feature.

Please fix this throughout the file, this pattern seems repeatead.

This method mixes two entirely separate functions, which should be
- moved to a new, a separate file `gps_ublox.c`
- split across 3 separate methods
-  `ubloxSetSbas` with `if/else`, a wrapper for the two methods:
`ubloxSetSbasM9orAbove`
`ubloxSetSbasM8orBelow`

Long, nested methods are not only hard to review, also hard/impossible to test.

Literally all [coding guidelines](https://google.github.io/styleguide/cppguide.html#Write_Short_Functions) discourage long methods, please make a conscious effort not to try to cram everything into a single function in a single file

Reply:
Significant refactoring is not within the scope of this PR, even though we agree entirely that it is needed.  Refactoring will be done only where it provides a fix for a functional problem, or arises naturally out of fixing a problem.

Reply:
Please refrain from telling us how to suck eggs.

Reply:
The `if/else` is still misplaced here as `ubloxSetSbas` is a combination of two distinct methods, `bloxSetSbasM9orAbove` and `ubloxSetSbasM8orBelow`. It's not asking for a "significant refactor", but comment on new code.
Not sure I understand your other comment either - you either don't know "how to suck eggs", or intentionally doing it the way that is not in line with standard engineering practices?

Reply:
The condition seems self-explanatory.  Modules M9 or higher respond to ValSet messages, so we send that kind of message to those modules; for other modules which will not respond to those messages, we send a different kind of message.  The if/else condition chooses which kind of message is sent to a particular kind of module.

It is true that this could be refactored as you describe, but the way it is written is easily understood.

Hence I don't understand what you mean by the if/else being misplaced.

Please refrain from slurs and speaking demeaningly to people who are trying their best with whatever skills they have; even if you are a far better coder than I am, it doesn't mean you should speak rudely.  

I would welcome refactoring PR's from you once this is merged.  As I said before, we agree that the code is in need of re-factoring.

Right now our focus is on fixing code problems that cause functional problems.  I welcome your assistance with that.  It requires testing and debugging.  That's what we value most at this point in time.

If there is a functional error with this code, then please explain what it is.